pkg/kv/kvserver/rditer/replica_data_iter.go

```diff
@@ -46,15 +46,25 @@ type ReplicaMVCCDataIterator struct {
 	reverse bool
 }
 
-// ReplicaEngineDataIterator is like ReplicaMVCCDataIterator, but iterates
-// using the general EngineKeys. It provides a subset of the engine.EngineIterator
-// interface.
+// ReplicaEngineDataIterator provides a complete iteration over all data in a
+// range, including system-local metadata and user data. The ranges KeyRange
+// slice specifies the key ranges which comprise the range's data.
+//
+// The iterator iterates over both point keys and range keys (i.e. MVCC range
+// tombstones), but in a somewhat peculiar order: for each key range, it first
+// iterates over all point keys in order, then over all range keys in order,
+// signalled via HasPointAndRange(). This allows efficient non-interleaved
+// iteration of point/range keys, and keeps them grouped by key range for
+// efficient Raft snapshot ingestion into a single SST per key range.
+//
+// TODO(erikgrinaker): Reconsider the above ordering/scheme for point/range
```

Comment:
Totally agree with TODO, we really need to think about what to do with this iterator. That would definitely violate principle of the least surprise for me.

Maybe we can have a visitor that would feed data in certain order to callbacks?

Reply:
In some sense, I suppose an iterator and a visitor are really the same pattern -- they emit some sequence of values. The issue here is really just that this iterator uses a different sequence than our other iterators. The sequence emitted by this iterator is exactly the sequence we want for the Raft snapshot protocol, but it is uncommon enough to be unintuitive/surprising.

I'm really leaning towards just getting rid of this iterator entirely, and just loop over the key ranges when we're sending the snapshot, creating regular engine iterators for each key range. We already have a Pebble snapshot available, and the code should be straightforward.

I'm going to push this through for now, to unblock other teams, but yeah -- let's revisit.

@tbg @sumeerbhola Any opinion on the above? For context:

https://github.com/cockroachdb/cockroach/pull/83747/files#diff-6abc68d5f62a6b7aa41e91cbb81a77ba965caa6d3b3f53c943c7a1fd597646bcR49-R58